Refactor features, implement new design for navigation & user menu

ryanclark opened the pull request:
This PR has a few architecture changes around features/routes, and implements the new navigation and user menu designs.

### Designs

Navigation - https://www.figma.com/file/v6GunK50D2VC7w7I2FBDNf/Teleport-Access-Management?node-id=121%3A6710&t=NDvvGzZNMMhw1oCp-0
User Menu - https://www.figma.com/file/Gpjs9vjhzUKF1GDbeG9JGE/Application-Design-System?node-id=997%3A2813&t=LE6vO4JJH2R3QCQu-0

### SVG Icons

I've added a bunch of the icons we're using in the new designs as SVG React components. These work way nicer than using font awesome (they're font awesome SVGs, but not using text for rendering). They look much more crisp than using the font alternative.

We should probably revisit the new `SVGIcons` components and include all of them at some point, but this PR is a good start to using them. They're named after their usage instead of their font awesome name, so it's more obvious which icon is to be used where.

### FeatureContext

The feature context seemed to only exist to pass through the features from `Teleport`/`TeleportE` through to the `Main` component, where it would then render the routes and register the features inside the monolithic `TeleportContext`.

I've changed this so the feature context only contains the features that are actually available to the authenticated user, and we can now use `useFeatures()` everywhere instead of pulling in `TeleportContext` and looking at `ctx.features`.

Features also now have a slightly different shape, as they define where they live in the new navigation (either under "Resources" or "Management", or not at all).

### Cleanup of `useX()`

This also cleans up a few non-helpful hook abstractions by putting their hooks into the main component that calls them. This simplifies `TopNav`, `Main` and `Clusters`.

It also means we can get rid of a few `ReturnType` values being used as the state of components. This is a major annoyance when trying to track down where different state values come from and the origins of props. We should strive to move away from this pattern.

## UI Changes

Here's a short video showing the new animations to go with the design

https://user-images.githubusercontent.com/7922109/214074373-b7a1ff48-5f04-4670-a6d3-dec45b9828ed.mov


### New navigation

<img width="1535" alt="image" src="https://user-images.githubusercontent.com/7922109/214073085-84a966c8-8f1b-4f82-bed7-69bc8fa618cf.png">

<img width="1535" alt="image" src="https://user-images.githubusercontent.com/7922109/214073108-eab3c4ae-629e-421b-a986-d77e901cd2d6.png">

<img width="277" alt="image" src="https://user-images.githubusercontent.com/7922109/214073138-6c651d63-9e63-4e0c-8e0a-76df07193397.png">

<img width="1535" alt="image" src="https://user-images.githubusercontent.com/7922109/214073187-bac1dc46-68c9-4519-94ac-bde34af0831f.png">

### User menu

<img width="381" alt="image" src="https://user-images.githubusercontent.com/7922109/214073256-3403fd58-63a1-4bcc-8b2e-749471542dfa.png">

### Other UX changes

The top bar is now 72px, matching the height of the design. I've overridden a few styles in our DarkSelect component to get the cluster selector inline with the designs also.

<img width="543" alt="image" src="https://user-images.githubusercontent.com/7922109/214073410-8638ee37-b8a8-4e8e-b656-7e1996e93ed4.png">

ryanclark committed:
Refactor features, implement new design for navigation & user menu

kimlisa reviewed (commented):
i really love the new navigation and had fun clicking around. couple of other things we need to do:

- add license headers to new files
- the `username` dropdown is also missing a hover highlight like we do for everything else 

optional nits:
- maybe it's just me that is bothered by this, but when I click on the `username` drop down, the animation seems different from the `management/resource` drop down, it seems to take longer and text slides
- prior we supported `keyboard tab` highlighting on the side navs, but we are missing it on the new nav (not that we had great support for it before so more like fyi)

but otherwise this is awesome stuff!

ryanclark commented on web/packages/design/src/SVGIcon/AccessRequests.tsx:
```diff
@@ -0,0 +1,20 @@
+      height={size}
+      fill={fill}
+    >
+      <path d="M7.5 7.875V5.625C7.5 5.00156 6.99844 4.5 6.375 4.5H4.5V0.375C4.5 0.16875 4.33125 0 4.125 0H3.375C3.16875 0 3 0.16875 3 0.375V4.5H1.125C0.501562 4.5 0 5.00156 0 5.625V7.875C0 8.49844 0.501562 9 1.125 9H3V23.625C3 23.8312 3.16875 24 3.375 24H4.125C4.33125 24 4.5 23.8312 4.5 23.625V9H6.375C6.99844 9 7.5 8.49844 7.5 7.875ZM6 7.5H1.5V6H6V7.5ZM13.125 15H11.25V0.375C11.25 0.16875 11.0813  ...
```
From font awesome, I copied the SVG from Figma

ryanclark commented on web/packages/teleport/src/features.tsx:
```diff
@@ -0,0 +1,548 @@
+export class FeatureApps implements TeleportFeature {
+  category = NavigationCategory.Resources;
+
+  topNavTitle = 'Applications';
```
For some reason we have topNavTitle that goes where the cluster selector is when browsing a non cluster specific URL.

We could probably just use `route.title` though right? And then we can get rid of `topNavTitle`.

ryanclark commented on web/packages/teleport/src/Recordings/__snapshots__/Recordings.story.test.tsx.snap:
```diff
@@ -254,63 +254,61 @@ exports[`rendering of Session Recordings 1`] = `
 
 .c3 .react-select__value-container {
-  padding: 0 8px;
+  padding: 0 16px;
```
I've scoped the changes to only the cluster selector at the top, so this is fixed

ryanclark commented on web/packages/teleport/src/Discover/Discover.tsx:
```diff
@@ -81,85 +57,23 @@ function DiscoverContent() {
-            return 'Are you sure you want to exit the “Add New Resource” workflow? You’ll have to start from the beginning next time.';
-          }}
-          when={selectedResource.shouldPrompt(currentStep)}
-        />
```
Added it back, we can discuss the wording in our sync tomorrow

ryanclark commented on web/packages/teleport/src/Discover/Navigation/StepItem.tsx:
```diff
@@ -114,9 +107,26 @@ const CheckedBullet = styled(Bullet)`
 `;
 
-const StepsContainer = styled<{ active: boolean }>(Text)`
+const StepsContainer = styled.div<{ active: boolean }>`
```
Fixed, nice spot

ryanclark commented:
> i really love the new navigation and had fun clicking around. couple of other things we need to do:

Cheers!

> * add license headers to new files

Done

> * the `username` dropdown is also missing a hover highlight like we do for everything else

The design doesn't have a hover state & Kenny was happy with just the text colour change, but I'll re-confirm with him

> optional nits:
> 
> * maybe it's just me that is bothered by this, but when I click on the `username` drop down, the animation seems different from the `management/resource` drop down, it seems to take longer and text slides

Yeah this is by design for actual popovers, the text also slides on the dropdown for resources but is a smaller change as it's coming from top down instead of right to left

> * prior we supported `keyboard tab` highlighting on the side navs, but we are missing it on the new nav (not that we had great support for it before so more like fyi)

I've added keyboard support - both tab and pressing the up/right arrows will navigate between links. Pressing right/left will open/close the Access Request dropdown when that's selected. I've also added keyboard navigation to the navigation selector between management/resources.
> but otherwise this is awesome stuff!

gzdunek reviewed (commented):
I added a few comments, but it's hard to review so many changes :(

I'm quite worried about the styles, it seems like values for font sizes, margins, colors, shadows are hardcoded in the components (instead of adding/taking them from `theme`).

Overall, good job!

ryanclark commented:
> I'm quite worried about the styles, it seems like values for font sizes, margins, colors, shadows are hardcoded in the components (instead of adding/taking them from `theme`).

Same, but our theme is so inconsistent that I figured we'll wait for the design system changes next quarter. I searched the codebase for some of the colours used from the design, and they only existed as hard-coded elsewhere.

ryanclark commented on web/packages/teleport/src/Clusters/Clusters.tsx:
```diff
@@ -14,42 +14,54 @@ See the License for the specific language governing permissions and
 
-export function Clusters(props: ReturnType<typeof useClusters>) {
-  const { clusters, enabledFeatures, initAttempt } = props;
+  const [clusters, setClusters] = useState([]);
+  const { attempt, run } = useAttempt();
```
I just moved the stuff from `useCluster()` into this component, I didn't really want to refactor it completely - only the parts that previously used `ctx.features`.

ryanclark commented on web/packages/teleport/src/Discover/Navigation/StepList.tsx:
```diff
@@ -16,25 +16,32 @@
-      {items}
-    </Box>
-  );
+  const items = [];
+
+  let index = props.index || 0;
+  for (const view of props.views) {
+    items.push(
+      <StepItem
+        key={index}
+        view={view}
+        currentStep={props.currentStep}
+        index={index}
+      />
+    );
+
+    index += 1;
+  }
```
Yeah I'm more a fan of `for .. of` loops to be honest, but I agree with the rename to `startIndex`

ryanclark commented on web/packages/teleport/src/Main/Main.tsx:
```diff
@@ -14,48 +14,77 @@ See the License for the specific language governing permissions and
+  const { attempt, setAttempt, run } = useAttempt('processing');
+
+  useEffect(() => {
+    if (ctx.storeUser.state) {
```
Not sure, this was the original code in `useMain` that I just moved into this function

ryanclark commented on web/packages/teleport/src/Main/Main.tsx:
```diff
@@ -14,48 +14,77 @@ See the License for the specific language governing permissions and
+
+  const features = useMemo(
+    () => props.features.filter(feature => feature.hasAccess(featureFlags)),
+    [featureFlags]
```
Good shout

ryanclark commented on web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx:
```diff
@@ -14,190 +14,238 @@
+  const initial =
+    username && username.length ? username.trim().charAt(0).toUpperCase() : '';
+
+  function handleClickOutside(event) {
```
Moved & added event

ryanclark replied:
Moved it back as ESLint complains, but I've added it as a dependency

gzdunek replied:
Hmm why is ESLint complaining? 
We already have functions in `useEffect` like here https://github.com/gravitational/teleport/blob/b23dcd680a07631f7935dc0a469d189b214977cc/web/packages/shared/hooks/useEscape.ts#L21

ryanclark replied:
```
  119:7  error  Move function declaration to function body root  no-inner-declarations
```

I guess ESLint doesn't recognise anonymous lambdas as function declarations, but I'm not a fan of defining functions as constant lambas - I'd rather just go for a named function block.

ryanclark commented:
> > maybe it's just me that is bothered by this, but when I click on the username drop down, the animation seems different from the management/resource drop down, it seems to take longer and text slides
> 
> I agree, I'd even say that animations for dropdowns are not needed at all. For me they are quite distracting :) I checked and GitHub/Slack/native dropdowns are not animated.

I guess it depends on where you look. Google Calendar/Material UI use animations to show dropdowns. You can check out our conversations regarding the animations & their fine tuning in #new-navigation in Slack

ryanclark commented on web/packages/teleport/src/Navigation/NavigationSwitcher.tsx:
```diff
@@ -0,0 +1,245 @@
+
+  const activeItem = props.items.find(item => item === props.value);
+
+  function handleClickOutside(event: MouseEvent) {
```
Good shout

ryanclark committed:
Move callbacks to use `useCallback`

ryanclark committed:
Update e ref

ryanclark committed:
Update e ref
